src/Java.Interop.Tools.JavaTypeSystem/Adapters/ManagedApiImporter.cs

```diff
@@ -0,0 +1,379 @@
+				: null;
+
+			if (ShouldSkipGeneric (type, non_generic_type, null))
+				return true;
+
+			return false;
+		}
+
+		static bool ShouldSkipGeneric (TypeDefinition? a, TypeDefinition? b, TypeDefinitionCache? cache)
+		{
+			if (a == null || b == null)
+				return false;
+			if (!a.ImplementsInterface ("Android.Runtime.IJavaObject", cache) || !b.ImplementsInterface ("Android.Runtime.IJavaObject", cache))
+				return false;
+
+			return GetRegisteredJavaTypeName (a) == GetRegisteredJavaTypeName (b);
+		}
+
+		static string? TypeReferenceToJavaType (TypeReference type)
+		{
+			var retval = GetRegisteredJavaName (type);
+
+			if (retval != null && type is GenericInstanceType generic) {
+				var parameters = generic.GenericArguments.Select (ga => GetRegisteredJavaName (ga.Resolve ())).ToArray ();
+
+				if (parameters.WhereNotNull ().Any ())
+					retval += $"<{string.Join (", ", parameters.WhereNotNull ())}>";
+			}
+
+			return retval;
+		}
+
+		static string? GetRegisteredJavaName (TypeReference type)
+		{
+			var td = type.Resolve ();
+
+			return GetRegisteredJavaTypeName (td);
+		}
+
+		static void ParseImplementedInterfaces (TypeDefinition type, JavaTypeModel model)
+		{
+			foreach (var iface_impl in type.Interfaces) {
+				var iface = iface_impl.InterfaceType;
+				var iface_def = iface.Resolve ();
+
+				if (iface_def is null || iface_def.IsNotPublic)
+					continue;
+
+				if (GetRegisterAttribute (iface_def.CustomAttributes) is CustomAttribute reg_attr) {
+					var jni = (string) reg_attr.ConstructorArguments [0].Value;
+					var name = jni.Replace ('/', '.').Replace ('$', '.');
+
+					model.Implements.Add (new JavaImplementsModel (name, name, $"L{jni};"));
+				}
+			}
+		}
+
+		static string GetBaseTypeJni (TypeDefinition type)
+		{
+			// Find a Java base type, ignoring generic types, if nothing else it will be Java.Lang.Object
+			TypeDefinition? base_type = type;
+
+			while (true) {
+				base_type = base_type.BaseType?.Resolve ();
+
+				if (base_type is null)
+					break;
+
+				// These are the base types for Java.Lang.Object and Java.Lang.Throwable
+				if (base_type.FullName == "System.Object" || base_type.FullName == "System.Exception")
+					return string.Empty;
+
+				if (base_type.HasGenericParameters || base_type.IsGenericInstance)
+					continue;
+
+				if (GetRegisterAttribute (base_type.CustomAttributes) is CustomAttribute reg_attr)
+					return (string) reg_attr.ConstructorArguments [0].Value;
+			}
+
+			return "java/lang/Object";
+		}
+
+		static CustomAttribute? GetObsoleteAttribute (Collection<CustomAttribute> attributes) =>
+			attributes.FirstOrDefault (a => a.AttributeType.FullNameCorrected () == "System.ObsoleteAttribute");
+
+		static CustomAttribute? GetRegisterAttribute (Collection<CustomAttribute> attributes) =>
+			attributes.FirstOrDefault (a => a.AttributeType.FullNameCorrected () == "Android.Runtime.RegisterAttribute");
+
+		static string? GetRegisteredJavaTypeName (TypeDefinition type)
```

Comment:
I think this could/should use [`JavaNativeTypeManager.GetJniTypeName()`](https://github.com/xamarin/java.interop/blob/855ecfa3a85ce74f40a1cc47e67b70aedd697dc3/src/Java.Interop.Tools.TypeNameMappings/Java.Interop.Tools.TypeNameMappings/JavaNativeTypeManager.cs#L458), as that should handle types without `[Register]` as well, or types with "Register-alternative" attributes.

Reply:
If it returns different things then I am scared to use it.  We don't know what knock-on effects that will have on **this** code.  It also does not seem to handle the special casing needed here like `"System.Collections.Generic.IList'1" => "java.util.List"`.